api/python/cellxgene_census/src/cellxgene_census/experimental/_embedding_search.py

```diff
@@ -0,0 +1,141 @@
+"""Nearest-neighbor search based on vector index of Census embeddings."""
+
+from contextlib import ExitStack
+from typing import Any, Dict, List, NamedTuple, Optional, Sequence
+
+import anndata as ad
+import numpy as np
+import numpy.typing as npt
+import pandas as pd
+import tiledb.vector_search as vs
+import tiledbsoma as soma
+
+from .._open import open_soma
+
+CENSUS_EMBEDDINGS_INDEX_URI_FSTR = (
+    "s3://cellxgene-contrib-public/contrib/cell-census/soma/{census_version}/indexes/{embedding_id}"
+)
+CENSUS_EMBEDDINGS_INDEX_REGION = "us-west-2"
+
+
+class NeighborObs(NamedTuple):
+    """Results of nearest-neighbor search for Census obs embeddings."""
+
+    distances: npt.NDArray[np.float32]
+    """
+    Distances to the nearest neighbors for each query obs embedding (q by k, where q is the number
+    of query embeddings and k is the desired number of neighbors). The distance metric is
+    implementation-dependent.
+    """
+
+    neighbor_ids: npt.NDArray[np.int64]
+    """
+    obs soma_joinid's of the nearest neighbors for each query embedding (q by k).
+    """
+
+
+def find_nearest_obs(
+    embedding_metadata: Dict[str, Any],
```

Comment:
Why does this use a different way to specify an embedding than [`get_embedding`](https://chanzuckerberg.github.io/cellxgene-census/_autosummary/cellxgene_census.experimental.get_embedding.html#cellxgene_census.experimental.get_embedding) does?

Reply:
`get_embedding` is a little low level in that it wants the full URI to the embeddings TileDB array, which isn't actually needed to find the index. `embedding_metadata` is the information returned from `get_embedding_metadata()` which seems like the appropriate level (especially in view of #1181 wherein we will actually put the relative URIs to the index arrays in there), although of course it'd be nice if it were more typesafe.

Reply:
@mlin can we change it to?

```
   embedding_name: str, 
   organism: str, 
   census_version: str
```

I think that provides an easier entry point to users and it aligns to [get_embedding_metadata_by_name](https://chanzuckerberg.github.io/cellxgene-census/_autosummary/cellxgene_census.experimental.get_embedding_metadata_by_name.html#cellxgene_census.experimental.get_embedding_metadata_by_name)

Reply:
This doesn't uniquely refer to an embedding, right? It refers to the newest embedding with this name?

I think we need to be able to non-ambiguously specify an embedding here since it's really important that the embedding being queried matches what we projected our data into.

Reply:
At the moment the combination of the three pieces does match to a unique embedding matrix, unless there is an edge-case I'm not catching.

Reply:
The documentation for `get_embedding_metadata_by_name` says that it can be ambiguous and returns the latest one.

I think the unique value for embeddings is in the `"id"` key of the embedding metadata, while `"embedding_name"` just says what method was used.

https://github.com/chanzuckerberg/cellxgene-census/blob/7612fb36e3504ea8886f855af0fd1cc8f67da10b/api/python/cellxgene_census/src/cellxgene_census/experimental/_embedding.py#L171-L172

Reply:
Right `"id"` is unique, but the IDs are `"cxg-czi-[\d+]"` or  `"cxg-contributed-[\d+]"` -- so no human-friendly.

The reason for  `If more embeddings match the query parameters, the most recent one will be returned.` is that we ended up having two versions of scVI for the 2023-12-15 Census version, the second effectively being a replacement of the first one (the Yosef lab recommended to that first needed some fixing).

So now refreshing my memory that one is actually the only edge-case we have, but it is intended for our users to use the latest version. I don't anticipate having to have multiple versions of the same embedding with the same name, unless one is the replacement of the other.